File: backend/apps/agents/manager/AgentLaunch.py
"""Agent run entry points for AgentManager: launch a new top-level run and the
invoke_agent helper (fork-and-send a sub-agent). The no-SDK mock fallback lives in MockAgent.
Split into a mixin to keep the manager file under the size ceiling; self.run_agent_loop /
self.sessions resolve across the MRO exactly as before."""

import logging
import os
from datetime import datetime
from uuid import uuid4

from typing import Dict, List, Optional

from typeguard import typechecked

from backend.apps.agents.core.models import (
    AgentConfig, AgentSession, Message, MessageBranch,
)
from backend.apps.agents.core.ws_manager import ws_manager
from backend.apps.settings.settings import load_settings
from backend.config.state_paths import home_state_dir, p_state_home
from backend.apps.agents.manager.session.session_store import load_session_data
from backend.apps.agents.manager.session.apply_context_window import apply_context_window
from backend.apps.agents.manager.session.workspace_git import (
    detect_git_identity,
    ensure_cwd_git_repo,
)
from backend.apps.agents.manager.prompt.tool_catalog import get_all_tool_names
from backend.apps.agents.manager.prompt.prompt_context import resolve_mode

logger = logging.getLogger(__name__)


from backend.apps.agents.manager.AgentManagerProtocol import AgentManagerProtocol


class AgentLaunch(AgentManagerProtocol):
    @typechecked
    async def launch_agent(self, config: AgentConfig) -> AgentSession:
        session_id = uuid4().hex
        # Phase log: a silent prompt drop used to leave NOTHING in the log to reason from. Each phase below names itself so a future one is diagnosable from a user's backend.log alone. Prompt length only, never its content.
        carried = len((config.initial_message.prompt if config.initial_message else "") or "")
        logger.info(
            "[launch %s] phase=start mode=%s model=%s target_dir=%s carries_prompt=%s prompt_chars=%d",
            session_id, config.mode, config.model, bool(config.target_directory), config.initial_message is not None, carried,
        )

        # Editing an existing App: when the user selected exactly one App card in App Builder mode, point the chat at that app's workspace so it edits in place. Without this the view-builder seed below fires (no target_directory) and registers a fresh empty "Untitled App" dupe.
        if (
            config.mode == "view-builder"
            and not config.target_directory
            and config.selected_app_output_ids
            and len(config.selected_app_output_ids) == 1
        ):
            from backend.apps.outputs.workspace_io import app_workspace_dir
            bound = app_workspace_dir(config.selected_app_output_ids[0])
            if bound:
                config.target_directory = bound

        mode_tools, _, mode_folder = resolve_mode(config.mode, get_all_tool_names)
        tools = mode_tools

        global_settings = load_settings()
        effective_cwd = (
            config.target_directory
            or mode_folder
            or global_settings.default_folder
            or p_state_home()
        )

        if config.mode in ("view-builder", "skill-builder") and not config.target_directory:
            effective_cwd = os.path.join(effective_cwd, session_id)

        os.makedirs(effective_cwd, exist_ok=True)
        logger.info("[launch %s] phase=cwd_resolved cwd=%s", session_id, effective_cwd)

        # Canvas-chat App Builder launch: when the user picks "App Builder" mode from the chat-input dropdown (no preexisting workspace, no target_directory passed in), the legacy code path only created an empty folder, so the agent could write files but the app never showed up in the Apps sidebar (no Output row, which is what the sidebar reads). Mirror the /workspace/seed endpoint's behavior here: seed the React template + register an Output row with workspace_id = session_id. Idempotent; safe if the session is ever re-launched with the same id.
        if config.mode == "view-builder" and not config.target_directory:
            try:
                from backend.apps.outputs.outputs import (
                    ensure_webapp_workspace_seeded_and_registered,
                )
                from backend.apps.outputs.workspace_io import load as load_output
                output_id = ensure_webapp_workspace_seeded_and_registered(
                    workspace_id=session_id,
                    folder=effective_cwd,
                    session_id=session_id,
                )
                logger.info("[launch %s] phase=workspace_seeded output_id=%s", session_id, output_id)
                if output_id:
                    # Broadcast the new row so the Apps sidebar lights up immediately, even before the user clicks into it. The row name is still the placeholder ("Untitled App") at this point; the post-session meta-sync below fires a second upsert with the real name once the agent has written meta.json.
                    try:
                        new_output = load_output(output_id)
                        await ws_manager.broadcast_global("agent:output_upserted", {
                            "output": new_output.model_dump(mode="json"),
                        })
                    except Exception:
                        logger.exception("post-seed output_upserted broadcast failed")
            except Exception:
                logger.exception(
                    "view-builder workspace seed/register failed; session will "
                    "still launch but the app may not appear in Apps sidebar"
                )

        # If the fallback chain landed on the state-home directory (no project dir, no default_folder set), re-route to a dedicated scratch workspace under <state_home>/.maestro/workspaces/<session_id>. This prevents us from writing .git/ (or anything else) into the real $HOME (or MAESTRO_STATE_HOME override) directly and gives the CLI's Agent tool a clean repo to do worktree isolation inside. Users with a default_folder or target_directory set keep whatever they configured.
        home = p_state_home()
        if os.path.abspath(effective_cwd) == os.path.abspath(home):
            effective_cwd = home_state_dir("workspaces", session_id)
            os.makedirs(effective_cwd, exist_ok=True)

        ensure_cwd_git_repo(effective_cwd, home)

        repo_url, branch_name = detect_git_identity(effective_cwd)

        session = AgentSession(
            id=session_id,
            name=config.name,
            provider=getattr(config, "provider", "anthropic"),
            model=config.model,
            mode=config.mode,
            system_prompt=config.system_prompt,
            allowed_tools=tools,
            max_turns=config.max_turns,
            cwd=effective_cwd,
            repo_url=repo_url,
            branch=branch_name,
            dashboard_id=config.dashboard_id,
            workflow_run_id=config.workflow_run_id,
            workflow_edit_id=config.workflow_edit_id,
            thinking_level=getattr(global_settings, "default_thinking_level", "auto"),
        )
        apply_context_window(session, global_settings)
        self.sessions[session_id] = session


        await ws_manager.send_to_session(session_id, "agent:status", {
            "session_id": session_id,
            "status": "running",
            "session": session.model_dump(mode="json"),
        })

        try:
            from backend.apps.service.analytics.client import track_agent_created
            track_agent_created(id=session.id, dashboard_id=session.dashboard_id)
        except Exception:
            pass

        # The session exists and is registered; delivery of any carried prompt happens next, in the endpoint, and logs its own phase.
        logger.info("[launch %s] phase=session_registered awaiting_prompt=%s", session_id, config.initial_message is not None)
        return session

    @typechecked
    async def invoke_agent(
        self,
        source_session_id: str,
        message: str,
        parent_session_id: Optional[str] = None,
        dashboard_id: Optional[str] = None,
    ) -> Dict:
        """Fork an existing session and send it a new message, returning the result."""
        source = self.sessions.get(source_session_id)
        if not source:
            data = load_session_data(source_session_id)
            if data is None:
                raise ValueError(f"Session {source_session_id} not found")
            source = AgentSession(**data)
            apply_context_window(source)

        source_name = source.name

        old_to_new_msg: Dict[str, str] = {}
        new_messages: List[Message] = []
        for msg in source.messages:
            new_id = uuid4().hex
            old_to_new_msg[msg.id] = new_id
            new_messages.append(Message(
                id=new_id,
                role=msg.role,
                content=msg.content,
                timestamp=msg.timestamp,
                branch_id=msg.branch_id,
                parent_id=old_to_new_msg.get(msg.parent_id) if msg.parent_id else None,
                # Sub-agents do NOT inherit parent's attached files. Each parent-message base64-expansion would re-fire in the sub-agent (cost explosion: a 25 MB PDF in parent + 5 InvokeAgent calls = 125 MB transmitted). The sub-agent receives the user's new message only; if it needs the file content, the parent message text from the prior turn already carries the model's summary.
                context_paths=None,
                attached_skills=msg.attached_skills,
                forced_tools=msg.forced_tools,
                images=msg.images,
            ))

        new_branches: Dict[str, MessageBranch] = {}
        for bid, branch in source.branches.items():
            new_branches[bid] = MessageBranch(
                id=bid,
                parent_branch_id=branch.parent_branch_id,
                fork_point_message_id=(
                    old_to_new_msg.get(branch.fork_point_message_id)
                    if branch.fork_point_message_id else None
                ),
                created_at=branch.created_at,
            )

        fork = AgentSession(
            id=uuid4().hex,
            name=f"{source_name} (invoked)",
            status="running",
            model=source.model,
            mode="invoked-agent",
            sdk_session_id=source.sdk_session_id,
            system_prompt=source.system_prompt,
            allowed_tools=list(source.allowed_tools),
            max_turns=source.max_turns or 25,
            cwd=source.cwd,
            created_at=datetime.now(),
            messages=new_messages,
            branches=new_branches,
            active_branch_id=source.active_branch_id,
            tool_group_meta=dict(source.tool_group_meta),
            dashboard_id=dashboard_id or source.dashboard_id,
            parent_session_id=parent_session_id,
        )
        apply_context_window(fork)

        self.sessions[fork.id] = fork

        await ws_manager.broadcast_global("agent:status", {
            "session_id": fork.id,
            "status": fork.status,
            "session": fork.model_dump(mode="json"),
        })

        user_msg = Message(
            role="user",
            content=message,
            branch_id=fork.active_branch_id,
        )
        fork.messages.append(user_msg)
        await ws_manager.send_to_session(fork.id, "agent:message", {
            "session_id": fork.id,
            "message": user_msg.model_dump(mode="json"),
        })

        await self.run_agent_loop(fork.id, message, fork_session=True)

        last_assistant = None
        for msg in reversed(fork.messages):
            if msg.role == "assistant":
                content = msg.content
                if isinstance(content, str):
                    last_assistant = content
                elif isinstance(content, list):
                    texts = [b.get("text", "") for b in content if isinstance(b, dict) and b.get("type") == "text"]
                    last_assistant = "\n".join(texts)
                else:
                    last_assistant = str(content)
                break

        return {
            "forked_session_id": fork.id,
            "source_name": source_name,
            "response": last_assistant or "No response from invoked agent.",
            "cost_usd": fork.cost_usd,
        }

File: engine/src/agents/manager/AgentLaunch.ts
// engine/src/agents/manager/AgentLaunch.ts -- AGT-5, a port of
// backend/apps/agents/manager/AgentLaunch.py: launchAgent (a new top-level run) and invokeAgent
// (fork-and-send a sub-agent).
//
// PORTS THE FIXED BEHAVIOUR, not the bug it replaced: AgentLaunch.py's workspace-location fallback
// resolves through `p_state_home()` (the override-aware state home, `pStateHome()` here), which
// was itself a same-day fix for a real bug (raw `os.path.expanduser("~")` leaked real `git init`'d
// workspaces into the developer's actual home when MAESTRO_STATE_HOME was overridden -- CTR-4's
// row in txm-status.md has the original discovery). The fix introduced a DELIBERATE split that
// this port carries over exactly: `pStateHome()` decides the workspace LOCATION (so
// MAESTRO_STATE_HOME isolation actually isolates), but `ensureCwdGitRepo`'s second argument is
// `realHome()` -- the REAL, un-overridden home -- because that argument is used ONLY to build the
// never-git-init-here guard (never touch $HOME, `/`, or $HOME's parent). Passing the state home
// there instead would narrow that guard to the state home, so a user pointing target_directory at
// their ACTUAL home while MAESTRO_STATE_HOME is overridden would slip past it. Getting this
// backwards reintroduces the exact bug the same-day fix closed -- see this file's own tests for a
// standing regression check.
//
// Scope cuts (each a stand-in for a not-yet-ported subsystem, documented at its own call site
// below): the view-builder workspace-seed-and-register side effect (outputs.outputs, SUB-5), the
// "editing an existing App" single-selected-app rebind (outputs.workspace_io, SUB-5), and the
// analytics track_agent_created call (backend/apps/service/analytics, not yet ported) are all DI
// seams defaulting to a no-op, matching the Python original's own try/except-and-continue shape
// around each of those calls -- a launch still succeeds identically without them, just without the
// extra UI/analytics side effect.

import { randomUUID } from 'node:crypto';
import { mkdirSync } from 'node:fs';
import { join, resolve } from 'node:path';
import type { AgentConfig, AgentSession, Message, MessageBranch } from '../core/models';
import { createAgentSession } from '../sessionFactory';
import { wsManager } from '../core/wsManager';
import { loadSettings } from '../../settings/store';
import { applyContextWindow } from './session/applyContextWindow';
import { detectGitIdentity, ensureCwdGitRepo } from './session/workspaceGit';
import { resolveMode } from './prompt/promptContext';
import { homeStateDir, pStateHome, realHome } from './statePaths';

function defaultGetAllToolNames(): string[] {
  return [];
}

export interface LaunchAgentDeps {
  getAllToolNames?: () => string[];
  loadSettings?: () => { settings: { default_folder: string | null; default_thinking_level: string } };
  /** Stand-in for outputs.outputs.ensure_webapp_workspace_seeded_and_registered (SUB-5, App
   * Builder). No default effect: a canvas-chat App Builder launch won't register an Output row
   * (so the Apps sidebar won't light up) until SUB-5 wires this, but the session itself still
   * launches and the agent can still write files into `effectiveCwd`. */
  ensureWebappWorkspaceSeededAndRegistered?: (args: { workspaceId: string; folder: string; sessionId: string }) => string | undefined;
  /** Stand-in for outputs.workspace_io.app_workspace_dir (SUB-5). No default: an "edit this one
   * selected App" launch falls through to the normal view-builder seed path instead of rebinding
   * to the existing app's workspace. */
  appWorkspaceDir?: (outputId: string) => string | undefined;
  /** Stand-in for the analytics.track_agent_created call. Default no-op. */
  trackAgentCreated?: (args: { id: string; dashboardId: string | null }) => void;
  newSessionId?: () => string;
}

/** Mirrors `launch_agent`: create and register a new session, resolving its working directory
 * through the same fallback chain (target_directory -> mode folder -> settings.default_folder ->
 * state home) the Python original uses, with the state-home-vs-real-home split described above. */
export async function launchAgent(sessions: Map<string, AgentSession>, config: AgentConfig, deps: LaunchAgentDeps = {}): Promise<AgentSession> {
  const getAllToolNames = deps.getAllToolNames ?? defaultGetAllToolNames;
  const loadSettingsFn = deps.loadSettings ?? (() => loadSettings());
  const sessionId = (deps.newSessionId ?? (() => randomUUID().replace(/-/g, '')))();

  // Editing an existing App: when exactly one App card was selected in App Builder mode with no
  // explicit target_directory, point the chat at that app's workspace so it edits in place.
  let targetDirectory = config.target_directory;
  if (config.mode === 'view-builder' && !targetDirectory && config.selected_app_output_ids?.length === 1 && deps.appWorkspaceDir) {
    const bound = deps.appWorkspaceDir(config.selected_app_output_ids[0]);
    if (bound) targetDirectory = bound;
  }

  const { tools: modeTools, default_folder: modeFolder } = resolveMode(config.mode, getAllToolNames);
  const globalSettings = loadSettingsFn().settings;

  let effectiveCwd = targetDirectory || modeFolder || globalSettings.default_folder || pStateHome();

  if ((config.mode === 'view-builder' || config.mode === 'skill-builder') && !targetDirectory) {
    effectiveCwd = join(effectiveCwd, sessionId);
  }

  mkdirSync(effectiveCwd, { recursive: true });

  // Canvas-chat App Builder launch: seed the React template + register an Output row so the app
  // shows up in the Apps sidebar immediately. See this file's header for the scope cut.
  if (config.mode === 'view-builder' && !targetDirectory && deps.ensureWebappWorkspaceSeededAndRegistered) {
    try {
      deps.ensureWebappWorkspaceSeededAndRegistered({ workspaceId: sessionId, folder: effectiveCwd, sessionId });
    } catch {
      // Best-effort, mirrors the Python original's logged-and-continue.
    }
  }

  // If the fallback chain landed on the state-home directory (no project dir, no default_folder
  // set), re-route to a dedicated scratch workspace under <state_home>/.maestro/workspaces/
  // <session_id>. This prevents writing .git/ (or anything else) into the real $HOME (or
  // MAESTRO_STATE_HOME override) directly.
  const home = pStateHome();
  if (resolve(effectiveCwd) === resolve(home)) {
    effectiveCwd = homeStateDir('workspaces', sessionId);
    mkdirSync(effectiveCwd, { recursive: true });
  }

  // Pass the REAL home, not the state home -- see this file's header for why. Handing it
  // pStateHome() would narrow the never-git-init-here guard to the state home, so a user pointing
  // target_directory at their actual home while MAESTRO_STATE_HOME is overridden would slip past it.
  ensureCwdGitRepo(effectiveCwd, realHome());

  const [repoUrl, branchName] = detectGitIdentity(effectiveCwd);

  const session = createAgentSession({
    id: sessionId,
    name: config.name,
    provider: config.provider || 'anthropic',
    model: config.model,
    mode: config.mode,
    system_prompt: config.system_prompt,
    allowed_tools: modeTools,
    max_turns: config.max_turns,
    cwd: effectiveCwd,
    repo_url: repoUrl ?? null,
    branch: branchName ?? null,
    dashboard_id: config.dashboard_id,
    workflow_run_id: config.workflow_run_id,
    workflow_edit_id: config.workflow_edit_id,
    thinking_level: (globalSettings.default_thinking_level as AgentSession['thinking_level']) ?? 'auto',
    created_at: new Date().toISOString(),
    branches: {},
  });
  applyContextWindow(session);
  sessions.set(sessionId, session);

  await wsManager.sendToSession(sessionId, 'agent:status', { session_id: sessionId, status: 'running', session } as never);

  try {
    deps.trackAgentCreated?.({ id: session.id, dashboardId: session.dashboard_id });
  } catch {
    // Best-effort, mirrors the Python original's try/except pass.
  }

  return session;
}

export interface InvokeAgentDeps {
  runAgentLoop?: (sessionId: string, prompt: string, opts: { forkSession: boolean }) => Promise<void>;
  newId?: () => string;
}

/** Mirrors `invoke_agent`: fork an existing session and send it a new message, returning the
 * result. `runAgentLoop` is DI'd -- wiring the fork to the REAL turn loop (AGT-4's TurnRunner, via
 * whichever ticket completes AgentManager.ts's non-mock path) is the caller's job, not this file's;
 * a default that throws would be wrong here because a caller that only wants the fork side effect
 * (no actual turn) is a legitimate use, so the default is a no-op that leaves the forked session
 * exactly as constructed (no assistant reply). */
export async function invokeAgent(
  sessions: Map<string, AgentSession>,
  sourceSessionId: string,
  message: string,
  opts: { parentSessionId?: string; dashboardId?: string } = {},
  deps: InvokeAgentDeps = {},
): Promise<{ forkedSessionId: string; sourceName: string; response: string; costUsd: number }> {
  const source = sessions.get(sourceSessionId);
  if (!source) throw new Error(`Session ${sourceSessionId} not found`);
  const sourceName = source.name;

  const newId = deps.newId ?? (() => randomUUID().replace(/-/g, ''));
  const oldToNewMsg = new Map<string, string>();
  const newMessages: Message[] = source.messages.map((msg) => {
    const newMsgId = newId();
    oldToNewMsg.set(msg.id, newMsgId);
    return {
      ...msg,
      id: newMsgId,
      parent_id: msg.parent_id ? (oldToNewMsg.get(msg.parent_id) ?? null) : null,
      // Sub-agents do NOT inherit the parent's attached files -- see the Python original's comment
      // for the cost-explosion reasoning.
      context_paths: null,
    };
  });

  const newBranches: Record<string, MessageBranch> = {};
  for (const [bid, branch] of Object.entries(source.branches)) {
    newBranches[bid] = {
      ...branch,
      fork_point_message_id: branch.fork_point_message_id ? (oldToNewMsg.get(branch.fork_point_message_id) ?? null) : null,
    };
  }

  const fork = createAgentSession({
    id: newId(),
    name: `${sourceName} (invoked)`,
    status: 'running',
    model: source.model,
    mode: 'invoked-agent',
    sdk_session_id: source.sdk_session_id,
    system_prompt: source.system_prompt,
    allowed_tools: [...source.allowed_tools],
    max_turns: source.max_turns || 25,
    cwd: source.cwd,
    created_at: new Date().toISOString(),
    messages: newMessages,
    branches: newBranches,
    active_branch_id: source.active_branch_id,
    tool_group_meta: { ...source.tool_group_meta },
    dashboard_id: opts.dashboardId ?? source.dashboard_id,
    parent_session_id: opts.parentSessionId ?? null,
  });
  applyContextWindow(fork);

  sessions.set(fork.id, fork);

  await wsManager.broadcastGlobal('agent:status', { session_id: fork.id, status: fork.status, session: fork } as never);

  const userMsg: Message = {
    id: newId(),
    role: 'user',
    content: message,
    timestamp: new Date().toISOString(),
    branch_id: fork.active_branch_id,
    parent_id: null,
    context_paths: null,
    attached_skills: null,
    forced_tools: null,
    images: null,
    hidden: false,
    client_message_id: null,
    elapsed_ms: null,
    tokens: null,
    tool_count: null,
    input_tokens: null,
  };
  fork.messages.push(userMsg);
  await wsManager.sendToSession(fork.id, 'agent:message', { session_id: fork.id, message: userMsg } as never);

  if (deps.runAgentLoop) {
    await deps.runAgentLoop(fork.id, message, { forkSession: true });
  }

  let lastAssistant: string | undefined;
  for (let i = fork.messages.length - 1; i >= 0; i--) {
    const msg = fork.messages[i];
    if (msg.role === 'assistant') {
      const content = msg.content;
      if (typeof content === 'string') {
        lastAssistant = content;
      } else if (Array.isArray(content)) {
        const texts = content
          .filter((b): b is { type: string; text?: string } => typeof b === 'object' && b !== null && (b as { type?: string }).type === 'text')
          .map((b) => b.text ?? '');
        lastAssistant = texts.join('\n');
      } else {
        lastAssistant = String(content);
      }
      break;
    }
  }

  return {
    forkedSessionId: fork.id,
    sourceName,
    response: lastAssistant || 'No response from invoked agent.',
    costUsd: fork.cost_usd,
  };
}
